src/services/meta_transaction_service.ts

```diff
@@ -0,0 +1,188 @@
+    ): Promise<CalculateMetaTransactionPriceResponse> {
+        const {
+            takerAddress,
+            sellAmount,
+            buyAmount,
+            buyTokenAddress,
+            sellTokenAddress,
+            slippagePercentage,
+            excludedSources,
+        } = params;
+
+        const assetSwapperOpts = {
+            ...ASSET_SWAPPER_MARKET_ORDERS_OPTS,
+            slippagePercentage,
+            bridgeSlippage: slippagePercentage,
+            excludedSources, // TODO(dave4506): overrides the excluded sources selected by chainId
+        };
+
+        let swapQuote;
+        if (sellAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketSellSwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                sellAmount,
+                assetSwapperOpts,
+            );
+        } else if (buyAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketBuySwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                buyAmount,
+                assetSwapperOpts,
+            );
+        } else {
+            throw new Error('sellAmount or buyAmount required');
+        }
+
+        const makerAssetAmount = swapQuote.bestCaseQuoteInfo.makerAssetAmount;
+        const totalTakerAssetAmount = swapQuote.bestCaseQuoteInfo.totalTakerAssetAmount;
+
+        const buyTokenDecimals = await serviceUtils.fetchTokenDecimalsIfRequiredAsync(
+            buyTokenAddress,
+            this._web3Wrapper,
+        );
+        const sellTokenDecimals = await serviceUtils.fetchTokenDecimalsIfRequiredAsync(
+            sellTokenAddress,
+            this._web3Wrapper,
+        );
+        const unitMakerAssetAmount = Web3Wrapper.toUnitAmount(makerAssetAmount, buyTokenDecimals);
+        const unitTakerAssetAMount = Web3Wrapper.toUnitAmount(totalTakerAssetAmount, sellTokenDecimals);
+        const price =
+            buyAmount === undefined
+                ? unitMakerAssetAmount.dividedBy(unitTakerAssetAMount).decimalPlaces(sellTokenDecimals)
+                : unitTakerAssetAMount.dividedBy(unitMakerAssetAmount).decimalPlaces(buyTokenDecimals);
+
+        const response: CalculateMetaTransactionPriceResponse = {
+            takerAddress,
+            sellAmount,
+            buyAmount,
+            price,
+            swapQuote,
+            sources: serviceUtils.convertSourceBreakdownToArray(swapQuote.sourceBreakdown),
+        };
+        return response;
+    }
+    public async calculateMetaTransactionQuoteAsync(
+        params: CalculateMetaTransactionQuoteParams,
+    ): Promise<GetMetaTransactionQuoteResponse> {
+        const { takerAddress, sellAmount, buyAmount, swapQuote, price } = await this.calculateMetaTransactionPriceAsync(
+            params,
+        );
+
+        const floatGasPrice = swapQuote.gasPrice;
+        const gasPrice = floatGasPrice.integerValue(BigNumber.ROUND_UP);
```

Comment:
This is in wei, so `8500000000` for example is 8.5 gwei. We want the equivalent of `Math.ceil(8500000000 * 1e-9) * 1e9`

Reply:
Good catch!